rlclientlib/dedup.cc

```diff
@@ -1,83 +1,288 @@
+        RETURN_ERROR_LS(nullptr, status, compression_error) << "Key not found while processing event into batch dictionary";
+      }
+      _used_objects.insert({ aid, 1 });
+      _size_estimate += sizeof(size_t) + content.size();
     }
     else
     {
-      ++it->second._count;
+      ++it->second;
     }
-    return hash;
   }
+  return error_code::success;
+}
 
-  bool dedup_dict::remove_object(generic_event::object_id_t oid)
-  {
-    auto it = _entries.find(oid);
-    if (it == _entries.end())
-      return false;
+size_t action_dict_builder::size() const
+{
+  return _size_estimate * _state.get_ewma_value();
+}
 
-    --it->second._count;
-    if (!it->second._count)
-      _entries.erase(it);
+int action_dict_builder::finalize(generic_event& evt, api_status* status)
+{
+  l::dedup_info_serializer ser;
+  generic_event::object_list_t action_ids;
+  const auto now = _state.get_time_provider() != nullptr ? _state.get_time_provider()->gmt_now() : timestamp();
+  std::vector<string_view> action_values;
 
-    return true;
-  }
+  RETURN_IF_FAIL(_state.get_all_values(_used_objects.begin(), _used_objects.end(), action_ids, action_values, status));
+  auto payload = ser.event(action_ids, action_values);
+
+  //remove used actions from the dictionary
+  RETURN_IF_FAIL(_state.remove_all_values(_used_objects.begin(), _used_objects.end(), status));
+
+  //compress the payload
+  size_t old_size = payload.size();
+  RETURN_IF_FAIL(_state.compress(payload, status));
+  size_t new_size = payload.size();
+
+  //update compression ratio estimator
+  _state.update_ewma(new_size / (float)old_size);
+
+  evt = generic_event(DEDUP_DICT_EVENT_ID, now, generic_event::payload_type_t::PayloadType_DedupInfo, std::move(payload));
+  return error_code::success;
+}
+
+template <typename event_t>
+struct dedup_collection_serializer
+{
+  using serializer_t = logger::fb_event_serializer<event_t>;
+  using buffer_t = utility::data_buffer;
+  using shared_state_t = dedup_state;
+
+  static int message_id() { return logger::message_type::fb_generic_event_collection; }
+
+  dedup_collection_serializer(buffer_t& buffer, content_encoding_enum content_encoding, shared_state_t& state)
+      : _dummy(0), _ser(buffer, content_encoding, _dummy), _state(state), _builder(state) {}
 
-  string_view dedup_dict::get_object(generic_event::object_id_t oid) const
+  int add(event_t& evt, api_status* status = nullptr)
   {
-    auto it = _entries.find(oid);
-    if (it == _entries.end())
-      return string_view();
-    return string_view(it->second._content.data(), it->second._length);
+    RETURN_IF_FAIL(_builder.add(evt.get_object_list(), status));
+    return _ser.add(evt, status);
+  }
+
+  uint64_t size() const {
+    return _ser.size() + _builder.size();
   }
 
-  int dedup_dict::transform_payload_and_add_objects(const char* payload, std::string& edited_payload, generic_event::object_list_t& object_ids, api_status* status)
+  int finalize(api_status* status)
```

Comment:
`status` to default to `nullptr`, same as all other method signatures with `status`?

Reply:
This usage of default values on internal APIs is wrong and we must stop doing it.
There is never a reason why we'd internally not pass a `status` objects. It only enables us to make mistakes.